src/lfo.rs

```diff
@@ -0,0 +1,150 @@
+use std::{f32::consts::PI};
+
+use crate::ring_buffer::RingBuffer;
+macro_rules! assert_close {
+    ($left:expr, $right:expr, $epsilon:expr) => {{
+        let (left, right, epsilon) = ($left, $right, $epsilon);
+        assert!(
+            (left - right).abs() <= epsilon,
+            "{} is not close to {} within an epsilon of {}",
+            left,
+            right,
+            epsilon
+        );
+    }};
+}
+
+pub struct LFO{
+    frequency: f32,
+    amplitude: f32,
+    waveType: WaveType,
+    buffer: RingBuffer<f32>,
+    sample_rate: f32,
+    phase_offset: f32
+}
+pub enum LfoParams{
+    Frequency,
+    Amplitude,
+}
+
+pub enum WaveType{
+    Sine,
+    Tri,
+    Saw,
+    Square,
+}
+
+impl LFO{
+    /// Creates a new 
+    pub fn new(sample_rate_: f32, waveType_: WaveType, frequency_: f32, amplitude_: f32)->Self{
+        let buff = RingBuffer::<f32>::new(sample_rate_ as usize);
+        let mut lfo = LFO{
+            waveType: waveType_,
+            frequency: frequency_,
+            amplitude: amplitude_,
+            sample_rate: sample_rate_,
+            buffer: buff,
+            phase_offset: 0.0
+        };
+        lfo.set_frequency(frequency_);
+        lfo.set_amplitude(amplitude_); 
+        lfo.fill_waveform();
+        lfo
+    }
+    /// Reads one sample at a time from the wavetable buffer
+    pub fn pop(&mut self) -> f32 {
+        //dbg!(self.phase_offset);
+        let val = self.buffer.get_frac(self.phase_offset);
+        self.phase_offset += self.frequency ;
+        val * self.amplitude
+    }
+    /// Gets sample from LFO wavetable at given offset
+    pub fn get(&self, phase_offset: f32) -> f32 {
+        self.buffer.get_frac(phase_offset) * self.amplitude
+    }
+    /// Sets LFO frequency and calculates new wavetable
+    pub fn set_frequency(&mut self, frequency: f32) {
+        self.frequency = frequency;
+    }
+    /// Sets amplitude of the LFO output
+    pub fn set_amplitude(&mut self, amplitude: f32){
+        self.amplitude = amplitude;
+    }
+    /// Sets LFO WaveType
+    pub fn set_wave_type(&mut self, waveType_: WaveType){
+        self.waveType = waveType_;
+        self.fill_waveform();
+    }
+    /// Refills the wavetable and resets the read pointer
+    pub fn reset(&mut self){
+        self.phase_offset = 0.0;
+        self.fill_waveform();
+    }
+    /// Returns the value of a given parameter
+    pub fn get_param(&mut self, param: LfoParams)->f32{
+        match param{
+            LfoParams::Frequency=> self.frequency,
+            LfoParams::Amplitude=>self.amplitude,
+        }
+    }
+    /// Fills internal wavetable buffer with current waveform
+    pub fn fill_waveform(&mut self){
+        let num_samples = self.buffer.capacity();
+
+        for i in 0..num_samples {
+            let phase = i as f32 / num_samples as f32;
+            let val = match self.waveType {
+                WaveType::Sine => (2.0 * PI * phase).sin(),
+                WaveType::Tri => {
```

Comment:
Surprising extra implementation on other kinds of waves for modulation, I wonder what would be the effect of these kinds of waves on vibrato.